File: src/boj/boj_1149.java
package boj;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class boj_1149 {
    static final int R = 0;
    static final int G = 1;
    static final int B = 2;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[][] arr = new int[n][3];
        for (int r = 0; r < n; r++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            for (int c = 0; c < 3; c++) {
                arr[r][c] = Integer.parseInt(st.nextToken());
            }
        }

        for (int i = 1; i < n; i++) {
            arr[i][R] += Math.min(arr[i - 1][G], arr[i - 1][B]);
            arr[i][G] += Math.min(arr[i - 1][R], arr[i - 1][B]);
            arr[i][B] += Math.min(arr[i - 1][R], arr[i - 1][G]);
        }
        System.out.println(Arrays.stream(arr[n - 1]).min().getAsInt());
        br.close();
    }
}

File: src/boj/boj_1149.kt
package boj

import java.io.BufferedReader
import java.io.InputStreamReader
import java.lang.Integer.min
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val n = readLine().toInt()

    val input = Array(n) {
        val stringTokenizer = StringTokenizer(readLine())
        IntArray(3) {
            stringTokenizer.nextToken().toInt()
        }
    }

    val cache = Array(n) { IntArray(3) }

    for (i in 0 until n) {
        for (j in 0 until 3) {
            cache[i][j] =  if (i == 0) {
                input[i][j]
            } else {
                input[i][j] + when (j) {
                    0 -> min(cache[i - 1][1], cache[i - 1][2])
                    1 -> min(cache[i - 1][0], cache[i - 1][2])
                    else -> min(cache[i - 1][0], cache[i - 1][1])
                }
            }
        }
    }

    println(min(cache[n - 1][0], min(cache[n - 1][1], cache[n - 1][2])))
    close()
}